File: src-1.x/BYTES.NET/Imaging/RGBImage.vb
'import internal namespace(s) required
Imports BYTES.NET.Imaging.API

Namespace Imaging

    Public Class RGBImage

        Implements IImage

#Region "private method(s)"

        Private _name As String = Guid.NewGuid.ToString

        Private _red As GraylevelImage = Nothing
        Private _green As GraylevelImage = Nothing
        Private _blue As GraylevelImage = Nothing

#End Region

#Region "public properties implementing the 'IImage' interface"

        Public Property Name As String Implements IImage.Name
            Get

                Return _name

            End Get
            Set(value As String)

                _name = value

            End Set
        End Property

#End Region

#Region "public new instance method(s)"

        ''' <summary>
        ''' default new instance method
        ''' </summary>
        ''' <param name="red"></param>
        ''' <param name="green"></param>
        ''' <param name="blue"></param>
        Public Sub New(ByRef red As GraylevelImage, ByRef green As GraylevelImage, ByRef blue As GraylevelImage)

            'set the variable(s)
            _red = red
            _green = green
            _blue = blue

        End Sub

#End Region

#Region "public method(s) implementing the 'IImage' interface"

        ''' <summary>
        ''' method returning the BitmapSource for displaying the image
        ''' </summary>
        ''' <returns></returns>
        Public Function GetBitmapSource() As BitmapSource Implements IImage.GetBitmapSource

            'get the pixel values
            Dim redValues As Integer(,) = _red.Values
            Dim greenValues As Integer(,) = _green.Values
            Dim blueValues As Integer(,) = _blue.Values

            'get the dimensions
            Dim width As Integer = _red.Width
            Dim Height As Integer = _red.Height

            'assemble the image data
            Dim stride As Integer = width * 3
            Dim numPaddingBytes As Integer = 0

            While True
                numPaddingBytes = stride Mod 4
                If (stride Mod 4) = 0 Then
                    Exit While
                End If

                stride += 1
                numPaddingBytes += 1
            End While

            Dim bits((stride - numPaddingBytes) * Height - 1) As Byte

            For col = 0 To width - 1
                For row = 0 To Height - 1

                    Dim pos As Integer = row * (stride - numPaddingBytes) + col * 3

                    bits(pos) = redValues(col, row)
                    bits(pos + 1) = greenValues(col, row)
                    bits(pos + 2) = blueValues(col, row)

                Next
            Next

            'return the output image
            Return BitmapSource.Create(width, Height, Nothing, Nothing, PixelFormats.Rgb24, Nothing, bits, stride)

        End Function

#End Region

    End Class

End Namespace

File: src-2023/BYTES.NET.Math/Imaging/RGBImage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using BYTES.NET.Math.Imaging.API;
using BYTES.NET.Persistance.API;

namespace BYTES.NET.Math.Imaging
{
    public class RGBImage : IImage
    {
        #region Private Fields

        private string _name = Guid.NewGuid().ToString();
        private GraylevelImage _red;
        private GraylevelImage _green;
        private GraylevelImage _blue;

        #endregion

        #region Public Properties Implementing IImage

        public string Name
        {
            get => _name;
            set => _name = value;
        }

        #endregion

        #region Constructors

        public RGBImage(ref GraylevelImage red, ref GraylevelImage green, ref GraylevelImage blue)
        {
            _red = red;
            _green = green;
            _blue = blue;
        }
        #endregion

        #region Public Methods Implementing IImage

        public byte[] GetRawImageData(out int width, out int height, out int stride)
        {
            width = _red.Width;
            height = _red.Height;

            stride = width * 3;
            int numPaddingBytes = 0;

            while (stride % 4 != 0) // Ensure 4-byte alignment
            {
                stride++;
                numPaddingBytes++;
            }

            byte[] bits = new byte[stride * height];

            for (int col = 0; col < width; col++)
            {
                for (int row = 0; row < height; row++)
                {
                    int pos = row * stride + col * 3;
                    bits[pos] = (byte)_red.Values[col + 1, row + 1];   // Red
                    bits[pos + 1] = (byte)_green.Values[col + 1, row + 1]; // Green
                    bits[pos + 2] = (byte)_blue.Values[col + 1, row + 1];  // Blue
                }
            }

            return bits;
        }

        public void UpdateFromIPersistable(IPersistable data)
        {
            throw new NotImplementedException();
        }

        public void loadImage(IImageParser parser)
        {
            throw new NotImplementedException();
        }
        #endregion
    }
}
